### ISSUE TITLE
Intermittent GCS -> GCS copy failures

### ISSUE THREAD

**Issue by JohnEmhoff** [rclone/rclone#3724]
#### What is the problem you are having with rclone?
Syncs between GCS <-> GCS fail intermittently with this error:
```
Failed to copy: googleapi: Error 413: Copy spanning locations and/or storage classes could not complete within 30 seconds. Please use the Rewrite method
```
Oddly enough, sometimes it'll eventually succeed on one of its retries, sometimes it won't and nothing gets copied. The single file in this transfer is pretty small (a couple megabytes).

#### What is your rclone version (output from `rclone version`)
```
rclone v1.50.1
- os/arch: linux/amd64
- go version: go1.13.4
```

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Ubuntu 18.04.2, 64 bit

####  Which cloud storage system are you using? (eg Google Drive)
Google cloud storage

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
`rclone --config /tmp/rclone.conf sync -vv --include \*.csv --transfers 1 --retries 10 gcs:source-bucket/path gcs:dest-bucket/path`

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
https://gist.github.com/JohnEmhoff/ea09ff7ccf4fd7aa1b89afb5fe4f3bca

Above is a log where it eventually succeeds on try five out of ten. Sometimes it goes ten tries and never succeeds.

**Comment by ncw** [comment]
The error seems to indicate that rclone should use this method instead: https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite

However it also says that this should only be needed for large objects so definitely not for a couple of megabytes.

Looking at the times in your logs it appears to be only waiting 8 seconds not 30 seconds for it to complete.

I'm tempted to say this is a problem at GCS - it might be worth searching current issues on the google bug tracker: https://issuetracker.google.com/ - I had a quick look and couldn't see anything.

I wonder if there is some timeout that we could be setting (didn't find anything in the docs).

Anyway 8 seconds to copy a 2MB file?  Something must be wrong!

**Comment by JohnEmhoff** [comment]
Oops, I missed a comma or something, the file is more like 200MB. I don't think that explains the errors but I guess it makes the timing make a little more sense.

**Comment by ncw** [comment]
I think even 200MB shouldn't take very long to server side copy.

I think to debug this further I'm going to need a failing HTTP transaction.

So if you could run rclone with `-vv --dump bodies --log-file rclone.log`

In that enormous log file you'll find the error text - this willl be part of an HTTP transaction.  What I need is the REQUEST part and the RESPONSE part.

If you like you can attach the log here or email it to me nick@craig-wood.com (with a link to the issue).

**Comment by JohnEmhoff** [comment]
Hmmm, I've been trying to reproduce and now the copies are happening essentially instantaneously, always on the first try. I guess we can chalk it up to a transient issue on the google side? Sorry to bug you...it was happening for more than a day so I figured it wasn't a server side issue.

**Comment by ncw** [comment]
> Hmmm, I've been trying to reproduce and now the copies are happening essentially instantaneously, always on the first try. I guess we can chalk it up to a transient issue on the google side? Sorry to bug you...it was happening for more than a day so I figured it wasn't a server side issue.

That is fine!  Glad it is working now. The problem looked like a GCS bug. You need a lot of evidence when presenting a bug to a cloud provider I've discovered ;-)

**Comment by theduke** [comment]
I'm still running into this issue.

Getting lots of errors on big files.

**Comment by ncw** [comment]
@theduke wrote:
> I'm still running into this issue.
> 
> Getting lots of errors on big files.

Are you getting the same error

    Error 413: Copy spanning locations and/or storage classes could not complete within 30 seconds. Please use the Rewrite method

How big are the files you are copying?

**Comment by theduke** [comment]
Yes, the same error, and the files are quite large (multiple Gigabytes).

**Comment by ncw** [comment]
I've re-implemented the Copy method using the Rewrite method as the error message suggests.

This works in my limited testing - can you give it a go @theduke ?

[v1.54.0-beta.4943.cc8538e0d.fix-3724-gcs-rewrite](https://beta.rclone.org/branch/fix-3724-gcs-rewrite/v1.54.0-beta.4943.cc8538e0d.fix-3724-gcs-rewrite/) on branch [fix-3724-gcs-rewrite](https://github.com/rclone/rclone/tree/fix-3724-gcs-rewrite) (uploaded in 15-30 mins)

**Comment by theduke** [comment]
That was quick!

Will try right away.

**Comment by theduke** [comment]
Still running, but no errors after +30 minutes while lot's were popping up before.

Thanks for the impressively quick response!

**Comment by theduke** [comment]
Can confirm a successful complete sync without errors.

**Comment by ncw** [comment]
Great!

It wasn't a complicated change once I'd looked at it a bit more :-)

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.54
